File: sistemas-lineares-lu/decomposicaoLU.py
import numpy as np
import scipy.linalg as la

n = 5 # Dimensão do sistema

np.random.seed(42)
A = np.random.rand(n, n)
H = la.hilbert(n)

e_vec = np.ones((n, 1))
b = A @ e_vec
c = H @ e_vec

P_A, L_A, U_A = la.lu(A)
P_H, L_H, U_H = la.lu(H)

y_A = la.solve_triangular(L_A, P_A.T @ b, lower=True)
x_A = la.solve_triangular(U_A, y_A)

y_H = la.solve_triangular(L_H, P_H.T @ c, lower=True)
x_H = la.solve_triangular(U_H, y_H)

erro_rel_A = np.linalg.norm(A @ x_A - b) / np.linalg.norm(x_A)
erro_rel_H = np.linalg.norm(H @ x_H - c) / np.linalg.norm(x_H)

print(f"Erro Relativo (Matriz A): {erro_rel_A:.4e}")
print(f"Erro Relativo (Matriz de Hilbert): {erro_rel_H:.4e}")

B = np.zeros((n, n))
I = np.eye(n)

for k in range(n):
    y_inv = la.solve_triangular(L_A, P_A.T @ I[:, k], lower=True)
    B[:, k] = la.solve_triangular(U_A, y_inv)

erro_inversa = np.linalg.norm(B - la.inv(A)) / np.linalg.norm(B)
erro_identidade = np.linalg.norm((A @ B) - I) / np.linalg.norm(I)

print(f"\nQualidade da Inversa Computada: {erro_inversa:.4e}")
print(f"Teste A * A^(-1) = I (Erro): {erro_identidade:.4e}")

File: sistemas-lineares-lu/decomposicaoLU.m
A = rand(n);
H = hilb(n);

e_vec = ones(n, 1);
b = A * e_vec;
c = H * e_vec;

[L_A, U_A, P_A] = lu(A);
[L_H, U_H, P_H] = lu(H);

y_A = L_A \ (P_A * b);
x_A = U_A \ y_A;

y_H = L_H \ (P_H * c);
x_H = U_H \ y_H;

erro_relativo_A = norm(A * x_A - b) / norm(x_A);
erro_relativo_H = norm(H * x_H - c) / norm(x_H);

fprintf('Erro Relativo na Matriz Aleatória: %e\n', erro_relativo_A);
fprintf('Erro Relativo na Matriz de Hilbert: %e\n', erro_relativo_H);

B = zeros(n);
I = eye(n);
for k = 1:n
    y_inv = L_A \ (P_A * I(:, k));
    B(:, k) = U_A \ y_inv;
end

erro_inversa = norm(B - inv(A)) / norm(B);
identidade_aprox = A * B;
erro_identidade = norm(identidade_aprox - I) / norm(I);

fprintf('Qualidade da Inversa (Norma do Resíduo): %e\n', erro_inversa);
fprintf('Teste A * A^(-1) = I (Erro): %e\n', erro_identidade);